### PR TITLE
feat: k8s cluster info using taints/tolerations incl. global ones

### PR DESCRIPTION
**Pull request by maxrussell**
## Description

Update to #6425 (which was reverted), but includes the global task container defaults' tolerations while deciding whether a node belongs to a resource pool. This fixes [an issue](https://app.circleci.com/pipelines/github/determined-ai/determined/36281/workflows/18c95a74-1bae-4617-a2a4-f44d98f4072f/jobs/1304145) detected by our end to end tests. Takes care of DET-9122.

## Test Plan

Same as #6425.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

DET-9122

### CODE REVIEW

File: master/internal/rm/kubernetesrm/pods.go [hunk 1]
```diff
@@ -934,99 +937,92 @@ func (p *pods) handleGetAgentsRequest(ctx *actor.Context) {
...
-					if decNumContainers := inf.NewDec(int64(numContainers),
-						0); i.Cmp(decNumContainers) < 0 {
-						container = &cproto.Container{
-							ID:    cproto.ID(id),
-							State: "RUNNING",
-						}
-					}
+		// We'll create a number of pseudo-containers in the summary equal to the number of
+		// running containers in this pool.
+		pseudoContainersAdded := 0
```
Comment (carolinaecalderon) [inline master/internal/rm/kubernetesrm/pods.go:996]: non-blocking question: I noticed that you initialized the Go variable here w/ shorthand and later with `var int ` -- for sake of consistency & my curiosity, are there different use cases?

File: master/internal/rm/kubernetesrm/pods.go [hunk 2]
```diff
@@ -1275,3 +1249,66 @@ func (p *pods) listConfigMapsInAllNamespaces(
...
+		if !taintTolerated(taint, tolerations) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func extractSlotInfo(node model.AgentSummary) (numSlots int, devType device.Type) {
+	var gpuSlots, cpuSlots int
```
Comment (carolinaecalderon) [inline master/internal/rm/kubernetesrm/pods.go:1285]: non-blocking question (cont'): this was the variable initialization I was referring to earlier -- are there any cases where gpuSlots/cpuSlots _wouldn't_ be zero (or be negative?) or why not initialize to 0? I did a quick search in the repo, and for cases like this (where the variable is incremented after some conditional inside a loop), the shorthand form is used most of the time.

**Review by carolinaecalderon — APPROVED** [review]
See comments for a couple style questions (non-blocking), but otherwise LGTM. I tested using [Eric's method](https://github.com/determined-ai/determined/pull/6425#issuecomment-1496515808).

**Comment by carolinaecalderon** [comment]
nit: just to make this cleaner, before merging, squash the 12 commits into 1.

File: master/internal/rm/kubernetesrm/pods.go (hunk 2 above)
Comment (maxrussell) [reply master/internal/rm/kubernetesrm/pods.go:1285]: Great question! In general, it doesn't matter. In Go, all variables start at their "zero value," which is `0` for numeric types, `""` for strings, and `nil` for pointers, slices, and maps. So this line is functionally equivalent to
```go
gpuSlots, cpuSlots := 0, 0
```

I wouldn't even say I have a strong reason for using one or the other in any particular circumstance other than when I want to be explicit about a variable's type (e.g. "the variable must be an `int` so I can return an int").

So definitely feel free to use whichever you think is more readable in any particular case.

**Pull request merged**